throttling/memory_gcra.go

```diff
@@ -0,0 +1,64 @@
+package throttling
+
+import (
+	"context"
+	"fmt"
+	"sync"
+	"time"
+
+	"github.com/throttled/throttled/v2"
+	"github.com/throttled/throttled/v2/store/memstore"
+
+	"github.com/rudderlabs/rudder-go-kit/cachettl"
+)
+
+const defaultMaxCASAttemptsLimit = 100
+
+type gcra struct {
```

Comment:
For this specific implementation we are effectively using 1+N stores:

1. One `cachettl` which holds all different GCRARateLimiterCtxs.
2. Multiple (throttled) memstores, one for every instance of GCRARateLimiterCtx that we are creating.

It should be possible to use our own implementation of `throttled.GCRAStoreCtx` e.g. by leveraging `cachettl` and avoid using throttled's own memstore. Or since each `GCRARateLimiterCtx`we are creating is handling a single key, we can make a store that always holds a singular item :)

Reply:
@atzoum what do you [think](https://github.com/rudderlabs/rudder-go-kit/pull/21/commits/007e36740ff20e3c372c4ee0e5346b559fcb9b39) of this approach?

The benchmark gives:

```
goos: linux
goarch: amd64
pkg: github.com/rudderlabs/rudder-go-kit/throttling
cpu: 12th Gen Intel(R) Core(TM) i9-12900K
BenchmarkInMemoryGCRA
BenchmarkInMemoryGCRA/one_unlimited_store_per_throttler
BenchmarkInMemoryGCRA/one_unlimited_store_per_throttler/single_key
BenchmarkInMemoryGCRA/one_unlimited_store_per_throttler/single_key-24         	 8378788	       141.1 ns/op	       0 B/op	       0 allocs/op
BenchmarkInMemoryGCRA/one_unlimited_store_per_throttler/multiple_keys
BenchmarkInMemoryGCRA/one_unlimited_store_per_throttler/multiple_keys-24      	 6042226	       197.6 ns/op	      46 B/op	       0 allocs/op
BenchmarkInMemoryGCRA/one_unlimited_store_for_all_throttlers
BenchmarkInMemoryGCRA/one_unlimited_store_for_all_throttlers/single_key
BenchmarkInMemoryGCRA/one_unlimited_store_for_all_throttlers/single_key-24    	 8469865	       141.0 ns/op	       0 B/op	       0 allocs/op
BenchmarkInMemoryGCRA/one_unlimited_store_for_all_throttlers/multiple_keys
BenchmarkInMemoryGCRA/one_unlimited_store_for_all_throttlers/multiple_keys-24 	 5978118	       223.6 ns/op	      24 B/op	       0 allocs/op
BenchmarkInMemoryGCRA/one_single_key_store_per_throttler
BenchmarkInMemoryGCRA/one_single_key_store_per_throttler/single_key
BenchmarkInMemoryGCRA/one_single_key_store_per_throttler/single_key-24        	 6954384	       169.8 ns/op	      16 B/op	       1 allocs/op
BenchmarkInMemoryGCRA/one_single_key_store_per_throttler/multiple_keys
BenchmarkInMemoryGCRA/one_single_key_store_per_throttler/multiple_keys-24     	 4416854	       285.8 ns/op	     102 B/op	       3 allocs/op
BenchmarkInMemoryGCRA/custom_store_per_throttler
BenchmarkInMemoryGCRA/custom_store_per_throttler/single_key
BenchmarkInMemoryGCRA/custom_store_per_throttler/single_key-24                	 9134334	       130.0 ns/op	       0 B/op	       0 allocs/op
BenchmarkInMemoryGCRA/custom_store_per_throttler/multiple_keys
BenchmarkInMemoryGCRA/custom_store_per_throttler/multiple_keys-24             	 7789651	       161.9 ns/op	      16 B/op	       0 allocs/op
PASS
ok  	github.com/rudderlabs/rudder-go-kit/throttling	21.661s
```

The last one (i.e. `custom_store_per_throttler`) is the new implementation.